### ISSUE TITLE
cmd/tailscaled: add Linux support for popping URLs?

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#4564]
On macOS and Windows, the Tailscale GUIs support opening URLs as requested by the control plane.

So if you need to authenticate to reach something, the browser opens and you do the necessary auth at/via Tailscale.

On Linux we don't have an official/common GUI client, but many people use desktop Linux with the running-as-root `tailscaled` process.

We could probably make it work: e.g. map packets (which tailscaled is routing anyway) to a UID to a dbus session to ... something.

This might be a terrible idea and should likely be a tailscaled knob, but....

/cc @danderson @maisem @twitchyliquid64 @Xe for brainstorming.

**Comment by danderson** [comment]
In general, the issue with linux is that each step of what you loosely suggest is racy, so you can end up with a TOCTOU (time of check vs. time of use) security vuln. Mapping packets to UID involves looking up what process owns a port (which is racy), mapping UID to login sessions is racy (although the impact there is the feature doesn't work, not a security issue), and so forth.

In a perfect world, we use this as an excuse to build a GUI for linux, and then this is trivial because the GUI is already running in the correct login context and can do the work, like on other platforms :)

**Comment by bradfitz** [comment]
If we only do it on TCP packets and we gate the TCP packet on the lookup, then the kernel state doesn't go away and the race is gone.

**Comment by twitchyliquid64** [comment]
Two terrible ideas, both of which rely on xdg-portals & keeping track of the UID from `tailscale up`:

1. Walk `/run` for run directories with the correct UID and standard naming pattern, then connect to the dbus socket therein with the standard naming pattern
2. Walk `/proc/*/environ` for processes with the correct UID, read out `XDG_RUNTIME_DIR` or `DBUS_SESSION_ADDRESS`, connect to that.

**Comment by twitchyliquid64** [comment]
As for `/proc/net/unix`, I dont think theres a way to map sockets to processes there.

**Comment by bradfitz** [comment]
I don't think we want to tie it to `tailscale up`. It should survive reboots where you might have a new `DBUS_SESSION_ADDRESS` or whatever (I assume it has some random numbers in it?)

**Comment by twitchyliquid64** [comment]
I'm suggesting only working out the UID from `tailscale up`, but doing the rest at runtime.

`DBUS_SESSION_ADDRESS` changes with each user session, but if we are very against whirring about `/proc`, a symlink to it should be hanging out at `/run/user/<uid>/bus`.

**Comment by bradfitz** [comment]
Ah. Well we already have tailscale up --operator-user. We can reuse that.
